### PR TITLE
Rewrite the backend using Valum

### PR DESCRIPTION
**Pull request by arteymix**
There's still some work to do, but I think it gives a good overview of what Valum can do.

I will probably integrate Template-GLib to remove all the inlined HTML someday.

Here's what should remain to do:

 - [x] produce proper `ETag` for caching
 - [x] finish deployment with Docker (or anything else we use)
 - [x] finish backward compatibility
    - [x] reimplement `search.php`
    - [x] reimplement `tooltip.php`
 - [x] test with sphinx
 - [ ] transition to nicer URIs? (e.g. `/gio-2.0/GLib.File` instead of ``gio-2.0/GLib.File/index.htm`)
 - [x] install Valum on the CI (will require Trusty for Python 3.4 for Meson)
 - [ ] read and render templates asynchronously
 - [ ] query the database asynchronously (see http://jan.kneschke.de/projects/mysql/async-mysql-queries-with-c-api/ or use some thread-based approach)
 - [x] implement a LRU cache to avoid reading from the filesystem

### CODE REVIEW

**Comment by arteymix** [comment]
Would it be reasonable to render search results server-side if the user does not have JavaScript? I think I could do something like that once I get everything working.

It's working without JavaScript quite nicely at this moment. I realized that the frontend is fetching some specific templates which are not served at this time.

Also, how comfortable would you be of serving this off Docker?

**Comment by lewisgoddard** [comment]
Docker is fine for local testing, although we tend to use Vagrant at @elementary Deployment tends to occur on bare Digital Ocean installs.

**Comment by arteymix** [comment]
Valum can be installed system-wide or used locally with Meson. I don't have Ubuntu packages yet, but I think people at elementary are good at this. If that works well, it will be ready for a final release.

I maintain a Docker image for deployment based on the latest Ubuntu: https://hub.docker.com/r/valum/valum/. It is really convenient, but I understand if you prefer to host it along existing apps.

We could have a Vagrantfile in this repository which replicate the hosting environment with Valum installed.

**Comment by arteymix** [comment]
It seems like the VM is using an old verison of libsoup, but Valum is has some fallback mode for that. In more recent versions, it will only warn for providing the `interface` property and use GIO SocketAddress API.

It's up to you all if you accept the branding for using Valum. I really don't mind removing it.

**Comment by arteymix** [comment]
It should be fully backward-compatible now.

There's probably a bug on the client side and lack of understanding this aspect prevents me from fixing it:

```
jquery.min.js:124 Uncaught DOMException: Failed to execute 'appendChild' on 'Node': Nodes of type '#document' may not be inserted inside nodes of type '#document-fragment'.(…)
```

I'll finish caching with `ETag` and look upstream with Meson to get the 0.36 working on Travis CI.

**Comment by arteymix** [comment]
Feel free to test it and give feedback. I'm pretty much done here. For potential contributors, I wrote a lot of documentation: http://docs.valum-framework.org/en/latest/

**Comment by arteymix** [comment]
There's still search and tooltip to reimplement :)

**Comment by arteymix** [comment]
@lewisgoddard How is the Sphinx database deployed? I think we should add this to the `Vagrantfile`.

**Comment by arteymix** [comment]
Ok, there's a layout for reimplementing `/tooltip.php` and `/search.php` features with libgda.

**Comment by lewisgoddard** [comment]
@arteymix with great difficulty? @btkostner has the details on Sphinx deployment.

**Comment by btkostner** [comment]
@arteymix I tried to get down the basics down in the [README.md](https://github.com/Valadoc/valadoc-org/blob/master/README.md#searching) when I figured it out. It shouldn't be too hard. The port doesn't need to be accessable from the outside, and you can just use a standard mysql interface to connect to it and query. Let me know if I can be of help!

**Comment by btkostner** [comment]
I don't have push permission to that branch (or I just suck at `git` which is a very valid reason), so here are some of the changes I made to the `Vagrantfile`.

- change `"ubuntu/trusty64"` to `"ubuntu/xenial64"` for more up to date packages
- add `sphinxsearch` and `xsltproc` to line 70 `apt-get install`
- Add the following lines after the valum build to generate the docs
```
    # Build valadoc files
    cd /home/vagrant/valadoc-org
    make
    make app
    make build-docs
```
- Add the following lines after the docs generate to setup and run sphinx in daemon mode
```
    # Setup sphinx search
    mkdir -p /home/vagrant/valadoc-org/sphinx/storage
    ./configgen ./valadoc.org
    indexer --config ./sphinx.conf --all
    searchd --config ./sphinx.conf
```

**Comment by benwaffle** [comment]
Since we're not using php anymore, it doesn't make sense for the URLs to be `/search.php` and `/tooltip.php`

**Comment by arteymix** [comment]
It now builds docs and generates Sphinx index on provisioning. I will probably finish reimplementing search and tooltip feature by this week-end :)

**Comment by arteymix** [comment]
@benwaffle I agree too, but I think it would be nice to have something drop-in first, because I don't really want to touch the frontend.

Then we'll have plenty of room for major improvements. I would really like to push forward some content negotiation for serving our docs in JSON and XML as well.

File: Makefile [hunk 1]
```diff
@@ -140,6 +145,6 @@ test-examples: valadoc-example-tester
 #
 
 serve: default build-docs
-	FWD_SEARCH=1 FWD_TOOLTIP=1 php -S localhost:7777 -t ./valadoc.org ./valadoc.org/router.php
```
Comment (KubaKaszycki) [inline Makefile]: Nice - removing PHP. Wonder why it was still there :laughing:

**Comment by arteymix** [comment]
It seems like libgda is using an unsupported feature from Sphinx MySQL frontend. I'll investigate and try to patch it.

**Comment by arteymix** [comment]
I opened a ticket on bugzilla for the MySQL issue: https://bugzilla.gnome.org/show_bug.cgi?id=776581

**Comment by arteymix** [comment]
I just figured the problem, gda checks the MySQL version with `select version()`, which returns an error with Sphinx implementation.

It should be working pretty soon now :)

**Comment by arteymix** [comment]
I think I'll rewrite it using MySQL C API directly instead, it will be much more cleaner considering that it does not involve more code. 

I think that Sphinx will require a specific libgda provider if we want to make it work correctly because they've implemented their own *flavour* of MySQL's language.

**Comment by arteymix** [comment]
Ok, the whole MySQL thing is fixed and now only querying Sphinx properly is remaining. I'm too tired now so maybe this week-end I'll finish that up.

**Comment by btkostner** [comment]
Sweet. Let me know if you need info about Sphinx querying. It should be rather straightforward though.

Also, [this repo](https://github.com/eustasy/vala-docs.org) might interest you ;)

**Comment by arteymix** [comment]
@btkostner Maybe you could help me at this point? I'm not particularly good at Sphinx, but I finished all the interpolation features for the database wrapper so it should not be too difficult from here on.

**Comment by arteymix** [comment]
Ok, we now have basic JSON serialization which is the default. The client still work because it passes `Accept: text/plain`.

It might be nice to have a `text/html` representation for JavaScript-disabled search.

**Comment by Bat41** [comment]
Would you like me to modify the front to use the JSON endpoint for search?

**Comment by arteymix** [comment]
@Bat41 I will probably postpone this feature until the moment we get some solid spec for the programmable API we will offer. The current JSON output is just of a proof-of-concept.

**Comment by Bat41** [comment]
@arteymix OK, let me know when you'll need help. 🙂 

By the way, the JavaScript code is quite ugly. I think we could rewrite it using ES6/Babel to get something nicer and more maintainable.

**Comment by btkostner** [comment]
I completely agree with the ES6 idea. We have an open issue about autoprefixed the css, so we are going to include a build step at some point, just not in this PR.

**Comment by arteymix** [comment]
I'm working on https://github.com/chergert/glrucache to have minimal LRU caching when we read documents from the filesystem.

**Comment by arteymix** [comment]
Nice, it all got merged! https://github.com/chergert/glrucache/pull/3

We now use a LRU cache for serving the documents instead of fetching them systematically off the disk. I measured a save of ~10ms, but I think it should be more significant when different resources are fetched, breaking the implicit cache pattern of the OS.

**Comment by arteymix** [comment]
I added a hidden input for the search page to include the package, which is less hackish than checking the current URL.

It's pretty much all functional with fixed `@` in symbols.

We have to be careful when merging because the main template is embedded in the code and I think there's been some good UI work already.

**Comment by benwaffle** [comment]
Can we add glrucache in-tree?

**Comment by benwaffle** [comment]
Also I'd like a dockerfile for development so I don't have to have sphinx installed. Vagrant uses virtualbox which is kinda heavy

**Comment by arteymix** [comment]
We can add a Dockerfile. I think that if we stick on the same distribution, we can have a common provision script and avoid duplicating all the initialization.

I did not introduce Meson to keep this as a drop-in replacement. If we do this, then it will make a lot of sense to use the compatible dependencies via subprojects.

**Comment by benwaffle** [comment]
Lets use meson subprojects for valum/vsgi/glrucache like this: https://github.com/benwaffle/pulse-flow/blob/master/subprojects/libgtkflow.wrap

**Comment by desiderantes** [comment]
Sad to see this incomplete:( Is it still viable?

**Comment by ZanderBrown** [comment]
Hopefully this can be finished, so much could be done with a JSON/XML API built on top of this

**Comment by arteymix** [comment]
@ZanderBrown I'm really busy right now with my research project and I truly hope I could put some time on this side-project by the end of the summer.

If anyone want to take the lead, I'll be glad to coordinate the efforts.

**Comment by ZanderBrown** [comment]
If we have a clear TODO i'm willing to try & find time for a couple items

**Comment by arteymix** [comment]
I'll upload a fully working Vagrant setup tomorrow, fix the merge conflicts and assess the remaining work to be done into a proper TODO list.

I think it's okay to use the coming VSGI 1.0 and Valum 0.4 which are already near stabilization.

**Comment by benwaffle** [comment]
https://github.com/eustasy/vala-docs.org - this generates an XML file containing all of the info needed. Would be good if the valum backend used that

**Comment by btkostner** [comment]
The more and more I think about this, I think the best route would be to have a Valum backend that uses the XML file generator to create an API only backend server.

Then having a frontend project that consumes the API just like every other client (builder, docs, etc.)

This keeps the backend all in vala. Easy to edit and maintain for vala devs. Also gives us more freedom with the frontend to offer some cool things like an pwa, and offline support. Maybe even an easier time including guides and less API-y like documentation.

**Comment by arteymix** [comment]
For Docker, I intended to setup a docker-compose environment to run the sphinx search engine and the HTTP backend separately. I don't know if that is consistent with your future hosting goals, but that could be a nice-to-have setup for scaling and targeting other hosting providers.

The Vagrant setup works for testing purposes, but I'm afraid I cannot put more time and efforts on this side project for now.

**Comment by btkostner** [comment]
Eventually down the line I would love to have everything just running in docker containers on kubernetes, but it's going to be a while before that happens. I was thinking as far as deployment goes, having CI create a docker image that contains all of the compiled vala code and docs as that time. That way if you need to go back in time and look at previous docs for some reason, it's just changing the docker image version. This also makes it a lot easier to deploy and pull down locally if you need.

**Comment by tintou** [comment]
I'll close it as there is no activity since more than a year, feel free to reopen when ready

**Pull request closed without merging**